pkg/commands/waiter.go

```diff
@@ -0,0 +1,183 @@
+package commands
+
+import (
+	"context"
+	"fmt"
+	"github.com/pivotal/kpack/pkg/apis/build/v1alpha1"
+	corev1alpha1 "github.com/pivotal/kpack/pkg/apis/core/v1alpha1"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	watchTools "k8s.io/client-go/tools/watch"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/apis/duck"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+	"knative.dev/pkg/kmeta"
+)
+
+type ResourceWaiter interface {
+	Wait(runtime.Object) error
+	BuilderWait(runtime.Object, int64, int64) error
+}
+
+func NewResourceWaiter(dc dynamic.Interface) ResourceWaiter {
+	return NewWaiter(dc)
+}
+
+type Waiter struct {
+	dynamicClient dynamic.Interface
+}
+
+func NewWaiter(dc dynamic.Interface) *Waiter {
+	return &Waiter{dynamicClient: dc}
+}
+
+func (i *Waiter) Wait(ob runtime.Object) error {
+	m, ok := ob.(kmeta.OwnerRefable)
+	if !ok {
+		return errors.New("unexpected type")
+	}
+	return i.wait(ob, hasResolved(m.GetObjectMeta().GetGeneration()))
+}
+
+// Used to ensure the builder has resolved to the expected stack and store (importing)
+func (i *Waiter) BuilderWait(ob runtime.Object, expectedStoreGeneration, expectedStackGeneration int64) error {
```

Comment:
It is a bit of a bummer that we need to have specific BuilderWait call. Could the Wait call somehow take an additional wait func or something that could just be used for builders?

Reply:
updated